conans/client/tools/win.py

```diff
@@ -63,24 +65,64 @@ def build_sln_command(settings, sln_path, targets=None, upgrade_project=True, bu
     return command
 
 
-def vs_installation_path(version):
-    if not hasattr(vs_installation_path, "_cached"):
-        vs_installation_path._cached = dict()
+def vs_installation_path(version, preference=None):
 
-    if version not in vs_installation_path._cached:
-        vs_path = None
+    vs_installation_path = None
+
+    if not preference:
+        env_prefs = [var.lstrip().rstrip() for var in get_env("CONAN_VS_INSTALLATION_PREFERENCE",
+                                                               list())]
+        if env_prefs:
+            preference = env_prefs
+        else:  # default values
+            preference = ["Enterprise", "Professional", "Community", "BuildTools"]
+
+    # Try with vswhere()
+    try:
         legacy_products = vswhere(legacy=True)
         all_products = vswhere(products=["*"])
-        products = legacy_products + all_products
+        products = list(legacy_products + all_products)
+    except ConanException:
+        products = None
 
-        for product in products:
-            if product["installationVersion"].startswith(("%d." % int(version))):
-                vs_path = product["installationPath"]
+    vs_paths = list()
 
-        # Remember to cache result
-        vs_installation_path._cached[version] = vs_path
+    if products:
+        # remove repeated products
+        seenProducts = list()
```

Comment:
seen_products

Reply:
Also, we tend to use [ ] in the codebase instead of list(). Not a big issue, but I'd prefer it for consistency.